scrabble.js

```diff
@@ -1,16 +1,209 @@
+  'y': { 'points':  4, 'qty':  2, },
+  'z': { 'points': 10, 'qty':  1, },
+}
 
 const Scrabble = {
+  validate(word) {
+    let valid = /^[A-Za-z]+$/;
+    let match = word.match(valid) && word.length <= 7 ? true : false;
+    return match
+  },
+
   score(word) {
+    let score = 0;
+    if (this.validate(word)) {
+      word = word.toLowerCase();
+      if (this.wordCheck(word)) {
+      for (let letter of word) {
+        let current_score = TILEBAG[letter].points;
+        score += current_score;
+      }
+      word.length === 7 ? score += 50 : '';
+    }
+      return score;
+    } else {
+      throw 'Word contains invalid character';
+    }
+  },
 
+  breakTie(incumbent, challenger) {
+    if (incumbent.length == 7) {
+      return incumbent
+    } else if (challenger.length == 7) {
+      return challenger
+    }
+    else if (challenger.length < incumbent.length) {
+      return challenger
+    } else {
+      return incumbent
+    }
   },
-  highestScoreFrom(arrayOfWords) {
 
+  highestScoreFrom(arrayOfWords) {
+    if (arrayOfWords.length === 0) {
+      throw 'No words to score';
+    } else if (arrayOfWords.length === 1) {
+      return arrayOfWords[0];
+    } else {
+      let max = arrayOfWords[0]
+      for (let word of arrayOfWords) {
+        let word_score = this.score(word);
+        let max_score = this.score(max);
+        if (max_score < word_score) {
+          max = word;
+        } else if (max_score === word_score) {
+          max = this.breakTie(max, word);
+        }
+      }
+      return max;
+    }
   },
+
+  wordCheck(word) {
+    return words.check(word);
+    // it is passing tests, but won't implement it because then all the provided tests will break
+  }
 };
 
+Scrabble.TileBag = class {
+  constructor() {
+    this.tiles = JSON.parse(JSON.stringify(TILEBAG));
+  }
+
+  tilesCount() {
+    let count = 0
+    Object.keys(this.tiles).forEach((letter) => {
+      count += this.tiles[letter].qty;
+    });
+    return count
+  }
+
+}
+
 Scrabble.Player = class {
+  constructor(name) {
+    if (name) {
+      this.name = name;
+      this.plays = [];
+      this.hand = [];
+      this.tileBag = new Scrabble.TileBag;
+    } else {
+      throw 'Player requires a name.'
+    }
+  }
 
-};
+  play(word) {
+    if (this.hasWon()) {
+      return false;
+    } else if (Scrabble.validate(word) && this.lettersInHand(word)) {
+      this.plays.push(word);
+      return (`${this.name} played the word ${word}`);
+    } else {
+      throw `${word} is not a valid play.`;
+    }
+  }
+
+  lettersInHand(word) {
+    let currentHand = this.hand;
+    let letters = {}
+    for (let tile of currentHand) {
+      letters[tile] ? letters[tile] += 1 : letters[tile] = 1;
+    }
+
+    let hasTiles = true
+    for (let letter of word) {
+      letters[letter] >= 1 ? letters[letter] -= 1 : hasTiles = false;
+    }
+    return hasTiles;
+  }
+
+  totalScore() {
+    let total = 0
+    if (this.plays.length === 0) {
```

Comment:
This will functionally happen with the loop you currently have in the `else`. When the `plays` variable is empty then it won't execute any of the iteration logic and it will return the value of 0.